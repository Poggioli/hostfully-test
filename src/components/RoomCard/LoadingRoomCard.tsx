import {
  Card
} from "../ui/card";
import { FC } from "react";
import { Skeleton } from "../ui/skeleton";


const LoadingRoomCard: FC = () => {
  return (
    <Card className="w-full">
      <Skeleton className="h-[200px] w-full" />
      <div className="p-3 flex flex-col gap-4">
        <Skeleton className="h-6 w-full max-w-[150px]" />
        <Skeleton className="h-14 w-full" />
        <Skeleton className="h-5 w-full max-w-[100px]" />

        <div className="flex justify-between gap-4">
          <Skeleton className="h-9 w-full max-w-[100px]" />
          <Skeleton className="h-9 w-full max-w-[100px]" />
        </div>
      </div>
    </Card>
  )
}

export { LoadingRoomCard };
