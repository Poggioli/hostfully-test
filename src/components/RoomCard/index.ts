export * from './LoadingRoomCard'
export * from './RoomCard'
