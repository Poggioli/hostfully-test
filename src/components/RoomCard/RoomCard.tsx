import { useBookingDrawer } from "@/components/BookingDrawer"
import { Button } from "@/components/ui/button"
import { Card, CardContent, CardDescription, CardFooter, CardHeader, CardTitle } from "@/components/ui/card"
import { Room } from "@/service/Room"
import { AspectRatio } from "@radix-ui/react-aspect-ratio"
import { FC } from "react"

export type RoomCardProps = Room

const RoomCard: FC<RoomCardProps> = ({ name, description, pricePerDay, photos, id }) => {

  const { onOpen } = useBookingDrawer()

  const onReserve = () => {
    onOpen(id)
  }

  return (
    <Card className="w-full">
      <AspectRatio ratio={16 / 9} className="rounded-t-md overflow-hidden">
        <img
          loading="lazy"
          src={photos[0]}
          alt={`Photo of ${name}`}
          className="rounded-t-md object-cover w-full h-full"
        />
      </AspectRatio>
      <CardHeader className="flex flex-col pb-3">
        <CardTitle>{name}</CardTitle>
        <CardDescription className="line-clamp-3">{description}</CardDescription>
      </CardHeader>

      <CardContent className="flex justify-end pb-4">
        <span>
          <span className="text-sm font-bold">
            ${pricePerDay}
          </span>
          {" "}night
        </span>
      </CardContent>

      <CardFooter className="flex justify-end">
        <Button onClick={onReserve}>Reserve</Button>
      </CardFooter>
    </Card>
  )
}

export { RoomCard }
