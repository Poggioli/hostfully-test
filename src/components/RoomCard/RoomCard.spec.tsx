import { act, fireEvent, render } from "@testing-library/react";
import { RoomCard } from "./RoomCard";

const mockOnOpen = jest.fn();
jest.mock("@/components/BookingDrawer", () => ({
  useBookingDrawer: () => ({
    onOpen: mockOnOpen,
  })
}));

describe('RoomCard', () => {
  it(`GIVEN a RoomCard
      WHEN render
      THEN should render all props the right way`, () => {
    const { baseElement } = render(
      <RoomCard
        name="room name 1"
        description="room description 1"
        pricePerDay={350}
        photos={['']}
        id="room-id-1"
        pricePerGuest={50}
      />);
    expect(baseElement).toMatchSnapshot();
  });

  it(`GIVEN a RoomCard
      WHEN click to in button "Reserve"
      THEN should call onOpen function from useBookingDrawer`, () => {

    const { getByRole } = render(
      <RoomCard
        name="room name 1"
        description="room description 1"
        pricePerDay={350}
        photos={['']}
        id="room-id-1"
        pricePerGuest={50}
      />);

    const reserveButton = getByRole('button', {
      name: 'Reserve'
    });

    act(() => fireEvent.click(reserveButton));

    expect(mockOnOpen).toHaveBeenCalledTimes(1);
    expect(mockOnOpen).toHaveBeenCalledWith("room-id-1");
  });
})