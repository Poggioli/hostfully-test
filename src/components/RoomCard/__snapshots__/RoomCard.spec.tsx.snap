// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RoomCard GIVEN a RoomCard
      WHEN render
      THEN should render all props the right way 1`] = `
<body>
  <div>
    <div
      class="rounded-xl border bg-card text-card-foreground shadow w-full"
    >
      <div
        data-radix-aspect-ratio-wrapper=""
        style="position: relative; width: 100%; padding-bottom: 56.25%;"
      >
        <div
          class="rounded-t-md overflow-hidden"
          style="position: absolute; top: 0px; right: 0px; bottom: 0px; left: 0px;"
        >
          <img
            alt="Photo of room name 1"
            class="rounded object-cover w-full h-full"
            loading="lazy"
            src=""
          />
        </div>
      </div>
      <div
        class="space-y-1.5 p-6 flex flex-col gap-4 pb-3"
      >
        <h3
          class="font-semibold leading-none tracking-tight"
        >
          room name 1
        </h3>
        <p
          class="text-sm text-muted-foreground line-clamp-3"
        >
          room description 1
        </p>
      </div>
      <div
        class="p-6 pt-0 flex justify-end pb-4"
      >
        <span>
          <span
            class="text-sm font-bold"
          >
            $
            350
          </span>
           
          night
        </span>
      </div>
      <div
        class="items-center p-6 pt-0 flex justify-end"
      >
        <button
          class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground shadow hover:bg-primary/90 h-9 px-4 py-2"
        >
          Reserve
        </button>
      </div>
    </div>
  </div>
</body>
`;
